### PR TITLE
Fix issue where queries can fail or omit OOO samples if OOO head compaction occurs between creating a querier and reading chunks

### PR DESCRIPTION
**Pull request by charleskorn**
This PR fixes two related issues:
* queries can fail with an `cannot populate chunk XXX from block 2ZBXFNYVVFDXFPGSB1CHFNYQTZ: not found` error if OOO head compaction occurs while the query is being evaluated
* queries can omit data from the OOO head if OOO head compaction occurs while the query is being evaluated

### `cannot populate chunk XXX from block 2ZBXFNYVVFDXFPGSB1CHFNYQTZ: not found` error

This issue occurs if events happen in this sequence:

1. Call `DB.ChunksQuerier()`, which returns a querier `q`
2. Call `q.Select()`, which returns a series set `ss`
3. Retrieve an iterator `it` for a series with samples in the OOO head block (ie. call `ss.At().Iterator(...)`)
4. Run OOO head compaction
5. Use `it` to read the chunks for the series

Eventually, `it.Next()` will return `false`, and `it.Err()` will report a `cannot populate chunk XXX from block 2ZBXFNYVVFDXFPGSB1CHFNYQTZ: not found` error.

`2ZBXFNYVVFDXFPGSB1CHFNYQTZ` is the ULID of the OOO head block, defined [here](https://github.com/prometheus/prometheus/blob/ab2a7bb74fa32f9199e3b9d4a19c000c304a37e9/tsdb/ooo_head.go#L140).

This happens because the garbage collection that occurs during OOO head compaction does not wait for pending reads to complete, and so the iterator created in step 3 has a reference to a chunk that is removed during compaction and can't be read in step 5.

### Omitting data

This issue occurs if events happen in this sequence:

1. Call `DB.ChunksQuerier()`, which returns a querier `q`
2. Call `q.Select()`, which returns a series set `ss`
3. Run OOO head compaction
4. Retrieve an iterator `it` for a series with samples that were in the OOO head before compaction (ie. call `ss.At().Iterator(...)`)
5. Use `it` to read the chunks for the series, none of which will include the samples that were in the OOO head before compaction

It also happens if steps 2 and 3 happen in the reverse order (ie. head compaction happens before the call to `q.Select()`).

In this case, the data that was in the OOO head before compaction is omitted, and no error is returned, so query results are incorrect.

This happens because the list of blocks on disk is captured in step 1, but by the time we get to step 4, which evaluates the list of chunks for the series, the chunk that was in the OOO head has been written to disk and garbage collected from the head, so it isn't returned from either source (disk or head).

### Root cause

I believe the reason for this is that OOO compaction doesn't check for pending reads that overlap with the chunks to be garbage collected, unlike what is done for in-order head compaction [here](https://github.com/prometheus/prometheus/blob/ab2a7bb74fa32f9199e3b9d4a19c000c304a37e9/tsdb/head.go#L1093). 

Under normal circumstances (ie. when `DB.Compact()` is used), OOO head compaction only occurs after in-order head compaction. Because of this, most of the time, this issue won't happen in practice because the time range of a query that includes OOO samples will also include time range of the in-order head and therefore be protected by the pending read check done before compacting the in-order head, or be protected by another query running at the same time that includes the in-order head. However, queries that are only for the time range of blocks on disk and OOO head samples (ie. do not overlap with the time range of the in-order head) would be susceptible to this issue.

---

I'd suggest reviewing each commit individually, the key commits are:
* https://github.com/prometheus/prometheus/commit/8898a5624d93144d53753d5d6099d84d963d01ae and https://github.com/prometheus/prometheus/commit/07c443f273f0f1cb4133d05cc6185f4ee0933b18 add tests that demonstrates the issue
* https://github.com/prometheus/prometheus/commit/cc238228c091030fa229f85232fe5e8e43818b7d adds a naive solution that uses the query range and chunk time range to block OOO garbage collection running while queries touching chunks that will be GCed are in progress, but this runs the risk of never allowing garbage collection to proceed
* https://github.com/prometheus/prometheus/commit/e3d12ed468954a3ef4165085d661656f9d19351c improves on the previous commit by using the last compacted OOO chunk reference instead to block garbage collection

I'd suggest reviewing this carefully: while this fix makes the test pass, I suspect there are some edge cases I haven't considered.

### CODE REVIEW

**Comment by jesusvazquez** [comment]
> 2ZBXFNYVVFDXFPGSB1CHFNYQTZ is the ULID of the OOO head block, defined [here](https://github.com/prometheus/prometheus/blob/ab2a7bb74fa32f9199e3b9d4a19c000c304a37e9/tsdb/ooo_head.go#L140).

A minor comment, recently @colega has made the head ULIDs readable so they are easy to differentiate form actual blocks https://github.com/prometheus/prometheus/pull/13100 This should bring a bit of clarity in the future if more issues of this kind arise 🙏

File: tsdb/ooo_head_read.go [hunk 1]
```diff
@@ -37,26 +37,29 @@ var _ IndexReader = &OOOHeadIndexReader{}
...
 }
 
-// The passed lastMmapRef tells upto what max m-map chunk that we can consider.
-// If it is 0, it means all chunks need to be considered.
-// If it is non-0, then the oooHeadChunk must not be considered.
-func (oh *OOOHeadIndexReader) series(ref storage.SeriesRef, builder *labels.ScratchBuilder, chks *[]chunks.Meta, lastMmapRef chunks.ChunkDiskMapperRef) error {
+// lastForbiddenMmapRef gives the last mmap chunk that may be being garbage collected and so
+// any chunk at or before this ref will not be considered. 0 disables this check.
+//
+// lastMmapRef tells upto what max m-map chunk that we can consider. If it is non-0, then
```
Comment (jesusvazquez) [inline tsdb/ooo_head_read.go:60]:
```suggestion
// maxMmapRef tells upto what max m-map chunk that we can consider. If it is non-0, then
```

File: tsdb/ooo_head_read.go [hunk 2]
```diff
@@ -37,26 +37,29 @@ var _ IndexReader = &OOOHeadIndexReader{}
...
 func (oh *OOOHeadIndexReader) Series(ref storage.SeriesRef, builder *labels.ScratchBuilder, chks *[]chunks.Meta) error {
-	return oh.series(ref, builder, chks, 0)
+	return oh.series(ref, builder, chks, oh.lastForbiddenMmapRef, 0)
 }
 
-// The passed lastMmapRef tells upto what max m-map chunk that we can consider.
-// If it is 0, it means all chunks need to be considered.
-// If it is non-0, then the oooHeadChunk must not be considered.
-func (oh *OOOHeadIndexReader) series(ref storage.SeriesRef, builder *labels.ScratchBuilder, chks *[]chunks.Meta, lastMmapRef chunks.ChunkDiskMapperRef) error {
+// lastForbiddenMmapRef gives the last mmap chunk that may be being garbage collected and so
```
Comment (jesusvazquez) [inline tsdb/ooo_head_read.go:57]: I'm wondering if this should be `minMmapRef`. In a way we are constraining the list of chunks that can be considered for the query and for the future reader min and max may sound more easier to understand. Then the rest of the comment provide the context explaining where do these constraints come from

File: tsdb/db.go [hunk 3]
```diff
@@ -1880,11 +1898,23 @@ func (db *DB) Querier(mint, maxt int64) (storage.Querier, error) {
 			blocks = append(blocks, b)
 		}
 	}
-	var inOrderHeadQuerier storage.Querier
+
+	blockQueriers := make([]storage.Querier, 0, len(blocks)+2) // +2 to allow for possible in-order and OOO head queriers
+
+	defer func() {
```
Comment (jesusvazquez) [inline tsdb/db.go:1904]: I like shifting the defer to the beginning 👍  I'm just slightly worried that the err gets shadowed, if you look below we're declaring an err on each of the if code blocks but this err actually refers to the last `for` where we are iterating the list of TSDB blocks... so either I'd colocate this defer there or declare the error up here.

**Review by jesusvazquez — COMMENTED** [review]
Thank you for your contribution @charleskorn 

I clarified in a comment above that ULIDs for the head blocks should now be more readable, hopefully providing a better UX to the administrator when this kind of bugs come up. I bet it was hard to find this...

I've given a first pass of feedback with some small nits, I'm fairly confident that this will work since your tests are comprehensive and nothing else broke. 

Unfortunately our prombench does not test the OOO head so we wont be able to catch any possible issues on CI, we won't know if something breaks until this runs in prod for some time, this scares me a bit. I'll make sure whomever is crafting the next release is aware about this PR in case issues comes up we know what to revert.

I've asked @fionaliao to also review the changes since she is working on OOO compaction and her feedback might be useful.

While you rebase + address my comments I'll read a bit more on isolation just to make sure I did not miss anything.

File: tsdb/db.go (hunk 3 above)
Comment (charleskorn) [reply tsdb/db.go:1904]: Given that there is no declaration of `err` before the `defer`, doesn't `err` in here refer to the returned error?

File: tsdb/ooo_head_read.go (hunk 1 above)
Comment (charleskorn) [reply tsdb/ooo_head_read.go:60]: Good catch, thanks 🙂

File: tsdb/ooo_head_read.go (hunk 2 above)
Comment (charleskorn) [reply tsdb/ooo_head_read.go:57]: I originally played with the idea of calling it `minMmapRef`, but I thought that would be confusing: `minMmapRef` to me sounds like it is the first chunk that can be considered (ie. it's inclusive), but it's the chunk before the first chunk that can be considered (ie. it's exclusive).

Maybe `lastGarbageCollectedMmapRef` would be clearer?

Comment (jesusvazquez) [reply tsdb/ooo_head_read.go:57]: Oh you are right. 

> Maybe lastGarbageCollectedMmapRef would be clearer?

Yeah lets do it.

File: tsdb/db.go (hunk 3 above)
Comment (jesusvazquez) [reply tsdb/db.go:1904]: Oh you are right! hadn't seen that 👍 

In this case I don't have anything to add then. Thank you

File: tsdb/ooo_head_read.go (hunk 2 above)
Comment (charleskorn) [reply tsdb/ooo_head_read.go]: Done, I've changed the name in 4a14a44.

**Comment by jesusvazquez** [comment]
@charleskorn there is a notable difference in memory usage in the benchmark
image

But I have seen the same behavior in a benchmark I'm running for another PR.

Could you please rebase your branch with the latest changes in main so I can run a benchmark that is more up to date? We need to check for regressions because this PR is changing a few things.

File: tsdb/ooo_head.go [hunk 4]
```diff
@@ -111,22 +112,27 @@ type OOORangeHead struct {
 	// the timerange of the query and having preexisting pointers to the first
 	// and last timestamp help with that.
 	mint, maxt int64
+
+	isoState *oooIsolationState
 }
 
-func NewOOORangeHead(head *Head, mint, maxt int64) *OOORangeHead {
+func NewOOORangeHead(head *Head, mint, maxt int64, minRef chunks.ChunkDiskMapperRef) *OOORangeHead {
+	isoState := head.oooIso.TrackReadAfter(minRef)
```
Comment (fionaliao) [inline tsdb/ooo_head.go:120]: We start tracking the read in `NewOOORangeHead()` and remove it when closing the chunk reader returned from `Chunks()`. If anything errors between `NewOOORangeHead()` and `Chunks()` then we'd never stop tracking this ref. 

This isn't actually a problem right now - nothing can actually error between `NewOOORangeHead()` and `Chunks()`. However, you have to go through a fair bit of code to verify this and it could be easy for a bug to be introduced so I think we should still protect against that case. One way to do that would be to close the `isoState` if `NewBlockQuerier()` errors.

**Comment by charleskorn** [comment]
> @charleskorn there is a notable difference in memory usage in the benchmark
>
> ...
> 
> But I have seen the same behavior in a benchmark I'm running for another PR.
> 
> Could you please rebase your branch with the latest changes in main so I can run a benchmark that is more up to date? We need to check for regressions because this PR is changing a few things.

Done, just rebased it now.

File: tsdb/ooo_head.go (hunk 4 above)
Comment (charleskorn) [reply tsdb/ooo_head.go:122]: Good idea, I've added this in abb21d6.

**Comment by charleskorn** [comment]
Looks like the CI failure is due to linting issues affecting unrelated files. https://github.com/prometheus/prometheus/pull/13178 will fix the linting issues.

**Comment by jesusvazquez** [comment]
Benchmarks results looking good 🎉

**Review by jesusvazquez — APPROVED** [review]
LGTM, thanks @charleskorn  for addressing this issue and improving the code in between.

File: tsdb/db.go [hunk 5]
```diff
@@ -1906,44 +1936,40 @@ func (db *DB) Querier(mint, maxt int64) (storage.Querier, error) {
...
-	var outOfOrderHeadQuerier storage.Querier
 	if overlapsClosedInterval(mint, maxt, db.head.MinOOOTime(), db.head.MaxOOOTime()) {
-		rh := NewOOORangeHead(db.head, mint, maxt)
+		rh := NewOOORangeHead(db.head, mint, maxt, db.lastGarbageCollectedMmapRef)
 		var err error
-		outOfOrderHeadQuerier, err = NewBlockQuerier(rh, mint, maxt)
+		outOfOrderHeadQuerier, err := NewBlockQuerier(rh, mint, maxt)
 		if err != nil {
+			// If NewBlockQuerier() failed, make sure to clean up the pending read created by NewOOORangeHead.
+			rh.isoState.Close()
```
Comment (bboreham) [inline tsdb/db.go:1951]: Do we need a similar Close in `blockChunkQuerierForRange`?  (Line 2030 in this PR)

Comment (charleskorn) [reply tsdb/db.go:1951]: Yes we do, good catch. Fixed in https://github.com/prometheus/prometheus/pull/14341.
